Foundation/NSNotificationQueue.swift

```diff
@@ -15,26 +23,105 @@ public struct NSNotificationCoalescing : OptionSetType {
 }
 
 public class NSNotificationQueue : NSObject {
-    
+
+    internal typealias NotificationQueueList = NSMutableArray
+    internal typealias NSNotificationListEntry = (NSNotification, [String]) // Notification ans list of modes the notification may be posted in.
+    internal typealias NSNotificationList = [NSNotificationListEntry] // The list of notifications to post
+
+    internal let notificationCenter: NSNotificationCenter
+    internal var asapList = NSNotificationList()
+    internal var idleList = NSNotificationList()
+
+    // The NSNotificationQueue instance is associated with current thread.
+    // The _notificationQueueList represents a list of notification queues related to the current thread.
+    private static var _notificationQueueList = NSThreadSpecific<NSMutableArray>()
+    internal static var notificationQueueList: NotificationQueueList {
+        return _notificationQueueList.get() {
+            return NSMutableArray()
+        }
+    }
+
+    // The default notification queue for the current thread.
+    private static var _defaultQueue = NSThreadSpecific<NSNotificationQueue>()
     public class func defaultQueue() -> NSNotificationQueue {
-        NSUnimplemented()
+        return _defaultQueue.get() {
+            return NSNotificationQueue(notificationCenter: NSNotificationCenter.defaultCenter())
+        }
     }
     
     public init(notificationCenter: NSNotificationCenter) {
-        NSUnimplemented()
+        self.notificationCenter = notificationCenter
+        super.init()
+        NSNotificationQueue.registerQueue(self)
     }
-    
+
+    deinit {
+        NSNotificationQueue.unregisterQueue(self)
+    }
+
     public func enqueueNotification(notification: NSNotification, postingStyle: NSPostingStyle) {
-        NSUnimplemented()
+        enqueueNotification(notification, postingStyle: postingStyle, coalesceMask: [.CoalescingOnName, .CoalescingOnSender], forModes: nil)
     }
 
     public func enqueueNotification(notification: NSNotification, postingStyle: NSPostingStyle, coalesceMask: NSNotificationCoalescing, forModes modes: [String]?) {
-        NSUnimplemented()
+        var runloopModes = [NSDefaultRunLoopMode]
+        if let modes = modes  {
+            runloopModes = modes
+        }
+
+        if !coalesceMask.isEmpty {
+            self.dequeueNotificationsMatching(notification, coalesceMask: coalesceMask)
+        }
+
+        switch postingStyle {
+        case .PostNow:
+            let currentMode = NSRunLoop.currentRunLoop().currentMode
+            if currentMode == nil || runloopModes.contains(currentMode!) {
+                self.notificationCenter.postNotification(notification)
+            }
+        case .PostASAP:
```

Comment:
if you want to do another that is fine, I was going to merge this since it looks like a great start and passes building/tests on linux